scipy/optimize/_linprog_highs.py

```diff
@@ -0,0 +1,411 @@
+        MODEL_STATUS_RDOVUB: (4, 'HiGHS Status Code 10: HighsModelStatusREACHED_DUAL_OBJECTIVE_VALUE_UPPER_BOUND'),
+        MODEL_STATUS_PRIMAL_INFEASIBLE: (2, "The problem is infeasible."),
+        MODEL_STATUS_PRIMAL_UNBOUNDED: (3, "The problem is unbounded."),
+        MODEL_STATUS_OPTIMAL: (0, "Optimization terminated successfully."),
+        MODEL_STATUS_REACHED_TIME_LIMIT: (1, "Time limit reached."),
+        MODEL_STATUS_REACHED_ITERATION_LIMIT: (1, "Iteration limit reached."),
+    }
+
+    c, A_ub, b_ub, A_eq, b_eq, bounds, x0 = lp
+
+    lb, ub = bounds.T.copy()                # separate bounds, copy->C-cntgs
+    # highs_wrapper solves LHS <= A*x <= RHS, not equality constraints
+    lhs_ub = -np.ones_like(b_ub)*np.inf     # LHS of UB constraints is -inf
+    rhs_ub = b_ub                           # RHS of UB constraints is b_ub
+    lhs_eq = b_eq                           # Equality constaint is inequality
+    rhs_eq = b_eq                           # constraint with LHS=RHS
+    lhs = np.concatenate((lhs_ub, lhs_eq))
+    rhs = np.concatenate((rhs_ub, rhs_eq))
+
+    if issparse(A_ub) or issparse(A_eq):
+        A = vstack((A_ub, A_eq))
+    else:
+        A = np.vstack((A_ub, A_eq))
+    A = csc_matrix(A)
+
+    options = {
+        'presolve': presolve,
+        'sense': 1,  # minimization
+        'solver': solver,
+        'time_limit': time_limit,
+        'message_level': message_level*disp,
+        'dual_feasibility_tolerance': dual_feasibility_tolerance,
+        'dual_objective_value_upper_bound': dual_objective_value_upper_bound,
+        'ipm_optimality_tolerance': ipm_optimality_tolerance,
+        'primal_feasibility_tolerance': primal_feasibility_tolerance,
+        'simplex_crash_strategy': simplex_crash_strategy,
+        'simplex_dual_edge_weight_strategy': simplex_dual_edge_weight_strategy,
+        'simplex_primal_edge_weight_strategy': simplex_primal_edge_weight_strategy,
+        'simplex_strategy': simplex_strategy,
+        'simplex_update_limit': simplex_update_limit,
+    }
+
+    options['ipm_iteration_limit'] = maxiter
+    options['simplex_iteration_limit'] = maxiter
+
+    # np.inf doesn't work; use very large constant
+    rhs = _replace_inf(rhs)
+    lhs = _replace_inf(lhs)
+    lb = _replace_inf(lb)
+    ub = _replace_inf(ub)
+
+    #from scipy.optimize._highs.mpswriter import mpswriter
+    #mpswriter(b'test.mps', c, A, lhs, rhs, lb, ub, np.array([], dtype=np.int32))
+
+    res = highs_wrapper(c, A.indptr, A.indices, A.data, lhs, rhs,
+                        lb, ub, options)
+
+    # If we get a MODEL_ERROR, match behavior of other linprog implementations
+    # and return infeasible status
+    if res['status'] == MODEL_STATUS_MODEL_ERROR:
+        res['status'] == MODEL_STATUS_PRIMAL_INFEASIBLE
+
+    # HiGHS represents constraints as lhs/rhs, so
+    # Ax + s = b => Ax = b - s
+    # and we need to split up s by A_ub and A_eq
+    slack = res.get('slack', None)
```

Comment:
We can cleanup the logic here a bit. Also None is the default return value from the key lookup so can be skipped.

```python
    if 'slack' in res:
        slack = res['slack']
        con, slack = slack[len(b_ub):], slack[:len(b_ub)]
    else:
        slack, con = None, None

    sol = {'x': np.array(res['x']) if 'x' in res else None,
           'slack': np.array(slack) if slack else None,
           # TODO: Add/test dual info like:
           # 'lambda': res.get('lambda', None),
           # 's': res.get('s', None),
           'fun': res.get('fun'),
           'con': np.array(con) if con else None,
           'status': statuses[res['status']][0],
           'success': res['status'] == MODEL_STATUS_OPTIMAL,
           'message': statuses[res['status']][1],
           'nit': res.get('simplex_nit', 0) or res.get('ipm_nit', 0),
           'crossover_nit': res.get('crossover_nit'),
           }
    return sol
```

Reply:
Incorporated suggestions